File: matrix/valid_sudoku.java
package matrix;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public class valid_sudoku {
    
}

class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, HashSet<Character>> rowSet = new HashMap<>();
        Map<Integer, HashSet<Character>> colSet = new HashMap<>();
        Map<String, HashSet<Character>> squareSet = new HashMap<>();

        for(int r = 0; r < 9; r++){
            for(int c = 0; c < 9; c++){
                char curChar = board[r][c];
                if(curChar == '.'){
                    continue;
                }
                String squareKey = ((r/3) + "," + (c/3));
                if(rowSet.containsKey(r) && rowSet.get(r).contains(curChar) ||
                colSet.containsKey(c) && colSet.get(c).contains(curChar) ||
                squareSet.containsKey(squareKey) && squareSet.get(squareKey).contains(curChar)){
                    return false;
                }

                rowSet.putIfAbsent(r, new HashSet<>());
                rowSet.get(r).add(curChar);

                colSet.putIfAbsent(c, new HashSet<>());
                colSet.get(c).add(curChar);

                squareSet.putIfAbsent(squareKey, new HashSet<>());
                squareSet.get(squareKey).add(curChar);
            }
        }
        return true;
    }
}

File: matrix/valid_sudoku.py
import collections
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rowSet = collections.defaultdict(set)
        colSet = collections.defaultdict(set)
        squareSet = collections.defaultdict(set)

        for r in range(9):
            for c in range(9):
                if board[r][c] == '.':
                    continue
                if(board[r][c] in rowSet[r] or board[r][c] in colSet[c] or
                board[r][c] in squareSet[(r//3, c//3)]):
                    return False
                
                rowSet[r].add(board[r][c])
                colSet[c].add(board[r][c])
                squareSet[(r//3, c//3)].add(board[r][c])

        return True
